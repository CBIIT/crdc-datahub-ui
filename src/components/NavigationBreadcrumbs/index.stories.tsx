import type { Meta, StoryObj } from "@storybook/react";
import NavigationBreadcrumbs, { NavigationBreadcrumbsProps } from "./index";

const meta: Meta<NavigationBreadcrumbsProps> = {
  title: "Miscellaneous / Navigation Breadcrumbs",
  component: NavigationBreadcrumbs,
  tags: ["autodocs"],
} satisfies Meta<NavigationBreadcrumbsProps>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    entries: [
      { label: "Home", to: "/" },
      { label: "Study Management", to: "/studies" },
      { label: "mock-study-uuid" },
    ],
  },
};

export const Truncation: Story = {
  args: {
    entries: [
      { label: "Home", to: "/" },
      { label: "Data Explorer", to: "/data-explorer" },
      {
        label:
          "lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt",
      },
    ],
  },
  decorators: [
    (Story) => (
      <div style={{ width: "350px" }}>
        <Story />
      </div>
    ),
  ],
};
