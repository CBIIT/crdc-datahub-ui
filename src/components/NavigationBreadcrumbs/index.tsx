import NavigateNextIcon from "@mui/icons-material/NavigateNext";
import { Breadcrumbs, styled, Typography } from "@mui/material";
import { FC, memo } from "react";
import { Link } from "react-router-dom";

const StyledBreadcrumbs = styled(Breadcrumbs)({
  textUnderlineOffset: "3px",
  width: "100%",
  "& .MuiBreadcrumbs-li:last-of-type": {
    minWidth: 0,
    flex: 1,
  },
  "& .MuiBreadcrumbs-separator": {
    color: "#71767A",
    marginLeft: "4px",
    marginRight: "4px",
  },
});

const StyledBreadcrumb = styled(Typography)({
  fontFamily: "Public Sans",
  fontWeight: 400,
  fontSize: "16px",
  maxWidth: "100%",
  overflow: "hidden",
  textOverflow: "ellipsis",
  whiteSpace: "nowrap",
});

const StyledLink = styled(Link)({
  color: "inherit",
});

export type BreadcrumbEntry = {
  /**
   * The label to display for the breadcrumb
   */
  label: string;
  /**
   * The URL to navigate to when the breadcrumb is clicked
   *
   * If omitted, the breadcrumb will visually appear as a static item
   */
  to?: string;
};

export type NavigationBreadcrumbsProps = {
  /**
   * The entries to display in the breadcrumbs
   */
  entries: Array<BreadcrumbEntry>;
};

/**
 * A component that provides a standard implementation of navigational breadcrumbs. Supports:
 *
 * - Linking to a route
 * - Native CSS truncation of labels
 *
 * You must provide a container for the component. By default, there is no external margin or padding.
 *
 * @returns The navigation breadcrumbs component
 */
const NavigationBreadcrumbs: FC<NavigationBreadcrumbsProps> = ({ entries }) => (
  <StyledBreadcrumbs
    separator={<NavigateNextIcon />}
    aria-label="navigation breadcrumb"
    data-testid="breadcrumb-container"
  >
    {entries.map(({ label, to }, index) => (
      <StyledBreadcrumb
        key={label}
        color={to ? "#005EA2" : "#1B1B1B"}
        data-testid={`breadcrumb-entry-${index}`}
      >
        {to ? <StyledLink to={to}>{label}</StyledLink> : label}
      </StyledBreadcrumb>
    ))}
  </StyledBreadcrumbs>
);

export default memo<NavigationBreadcrumbsProps>(NavigationBreadcrumbs);
