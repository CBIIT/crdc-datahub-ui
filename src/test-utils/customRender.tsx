import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import theme from "../theme/testTheme";

const AllTheProviders = ({ children }) => <ThemeProvider theme={theme}>{children}</ThemeProvider>;

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
