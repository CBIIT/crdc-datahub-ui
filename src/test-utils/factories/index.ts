export * from "./ApprovedStudyFactory";
export * from "./BatchFactory";
export * from "./SubmissionStatisticFactory";
export * from "./UserFactory";
