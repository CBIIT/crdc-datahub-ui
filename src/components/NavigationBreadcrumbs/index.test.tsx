import { MemoryRouter } from "react-router-dom";
import { axe } from "vitest-axe";

import { render } from "../../test-utils";

import NavigationBreadcrumbs, { BreadcrumbEntry } from "./index";

describe("Accessibility", () => {
  it("should have no accessibility violations", async () => {
    const entries: BreadcrumbEntry[] = [
      { label: "link item", to: "/link-item" },
      { label: "non-link" },
    ];

    const { container } = render(<NavigationBreadcrumbs entries={entries} />, {
      wrapper: MemoryRouter,
    });

    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("Basic Functionality", () => {
  it("should render without crashing", () => {
    expect(() =>
      render(<NavigationBreadcrumbs entries={[]} />, {
        wrapper: MemoryRouter,
      })
    ).not.toThrow();
  });

  it("should render the correct number of breadcrumbs", () => {
    const entries: BreadcrumbEntry[] = [
      { label: "Home", to: "/" },
      { label: "Data Explorer", to: "/data-explorer" },
      { label: "Study View", to: "/data-explorer/study-view" },
      { label: "Metadata Details", to: "/data-explorer/study-view/metadata-details" },
    ];

    const { getAllByTestId } = render(<NavigationBreadcrumbs entries={entries} />, {
      wrapper: MemoryRouter,
    });

    expect(getAllByTestId(/breadcrumb-entry/i)).toHaveLength(4);
  });

  it("should support entries that are not links", async () => {
    const entries: BreadcrumbEntry[] = [
      { label: "Home" },
      { label: "Products" },
      { label: "Electronics" },
    ];

    const { queryAllByRole } = render(<NavigationBreadcrumbs entries={entries} />, {
      wrapper: MemoryRouter,
    });

    expect(queryAllByRole("link")).toHaveLength(0);
  });

  it("should style links correctly", () => {
    const entries: BreadcrumbEntry[] = [{ label: "Some Page", to: "/some-page" }];

    const { getByTestId } = render(<NavigationBreadcrumbs entries={entries} />, {
      wrapper: MemoryRouter,
    });

    expect(getByTestId("breadcrumb-entry-0")).toHaveStyle("color: #005EA2");
  });

  it("should style non-link items correctly", () => {
    const entries: BreadcrumbEntry[] = [{ label: "Non-Link Item" }];

    const { getByTestId } = render(<NavigationBreadcrumbs entries={entries} />, {
      wrapper: MemoryRouter,
    });

    expect(getByTestId("breadcrumb-entry-0")).toHaveStyle("color: #1B1B1B");
  });

  it("should support rerendering with different entries", () => {
    const initialEntries: BreadcrumbEntry[] = [{ label: "Initial" }];
    const newEntries: BreadcrumbEntry[] = [{ label: "Updated" }];

    const { rerender, getByTestId } = render(<NavigationBreadcrumbs entries={initialEntries} />, {
      wrapper: MemoryRouter,
    });

    expect(getByTestId("breadcrumb-entry-0")).toHaveTextContent("Initial");

    rerender(<NavigationBreadcrumbs entries={newEntries} />);

    expect(getByTestId("breadcrumb-entry-0")).toHaveTextContent("Updated");
  });
});

describe("Snapshots", () => {
  it("should match the snapshot", () => {
    const entries: BreadcrumbEntry[] = [
      { label: "Home", to: "/" },
      { label: "Data Explorer", to: "/data-explorer" },
      { label: "Study View", to: "/data-explorer/study-view" },
      { label: "Metadata Details", to: "/data-explorer/study-view/metadata-details" },
    ];

    const { container } = render(<NavigationBreadcrumbs entries={entries} />, {
      wrapper: MemoryRouter,
    });

    expect(container).toMatchSnapshot();
  });
});
