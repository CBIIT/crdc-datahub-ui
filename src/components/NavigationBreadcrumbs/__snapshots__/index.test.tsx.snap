// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshots > should match the snapshot 1`] = `
<div>
  <nav
    aria-label="navigation breadcrumb"
    class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-sm3bj9-MuiTypography-root-MuiBreadcrumbs-root"
    data-testid="breadcrumb-container"
  >
    <ol
      class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
    >
      <li
        class="MuiBreadcrumbs-li"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-utvzy7-MuiTypography-root"
          data-testid="breadcrumb-entry-0"
        >
          <a
            class="css-q2y3yl"
            href="/"
          >
            Home
          </a>
        </p>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="NavigateNextIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-utvzy7-MuiTypography-root"
          data-testid="breadcrumb-entry-1"
        >
          <a
            class="css-q2y3yl"
            href="/data-explorer"
          >
            Data Explorer
          </a>
        </p>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="NavigateNextIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-utvzy7-MuiTypography-root"
          data-testid="breadcrumb-entry-2"
        >
          <a
            class="css-q2y3yl"
            href="/data-explorer/study-view"
          >
            Study View
          </a>
        </p>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="NavigateNextIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-utvzy7-MuiTypography-root"
          data-testid="breadcrumb-entry-3"
        >
          <a
            class="css-q2y3yl"
            href="/data-explorer/study-view/metadata-details"
          >
            Metadata Details
          </a>
        </p>
      </li>
    </ol>
  </nav>
</div>
`;
